File: modulo08-arquitetura-de-sistemas-com-ia/modulo-02-single-agent/agent-components-demo-java/src/test/java/com/trialforge/agentcomponents/AcaoGateTest.java
package com.trialforge.agentcomponents;

import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Cobre o mesmo cenario de testarGate() em agent-components-demo.js / .py:
 * acao sem gate executa direto; acao com gate nunca chega a executar.
 */
class AcaoGateTest {

    @Test
    void acaoSemGate_executaDireto() {
        AcaoGate.ResultadoAcao resultado =
                AcaoGate.executarOuGatear(new AcaoGate.AcaoProposta("x", false, () -> "feito"));

        assertThat(resultado.status()).isEqualTo("executada");
        assertThat(resultado.resultado()).isEqualTo("feito");
        assertThat(resultado.mensagem()).isNull();
    }

    @Test
    void acaoComGate_nuncaChegaAExecutar() {
        boolean[] executou = {false};
        AcaoGate.ResultadoAcao resultado = AcaoGate.executarOuGatear(new AcaoGate.AcaoProposta("y", true, () -> {
            executou[0] = true;
            return "nunca deveria rodar";
        }));

        assertThat(resultado.status()).isEqualTo("aguardando_aprovacao");
        assertThat(resultado.resultado()).isNull();
        assertThat(resultado.mensagem()).contains("y").contains("Approval Gate");
        assertThat(executou[0]).isFalse();
    }
}

File: modulo08-arquitetura-de-sistemas-com-ia/modulo-02-single-agent/agent-components-demo-go/internal/agente/acaogate_test.go
package agente

import "testing"

// Cobre o mesmo cenario de testarGate() em agent-components-demo.js / .py:
// acao sem gate executa direto; acao com gate nunca chega a executar.

func TestExecutarOuGatear_AcaoSemGate_ExecutaDireto(t *testing.T) {
	resultado := ExecutarOuGatear(AcaoProposta{
		Tipo: "x", RequerAprovacao: false, Executar: func() string { return "feito" },
	})

	if resultado.Status != "executada" {
		t.Errorf("esperava status=executada, obteve %q", resultado.Status)
	}
	if resultado.Resultado != "feito" {
		t.Errorf("esperava resultado=feito, obteve %q", resultado.Resultado)
	}
	if resultado.Mensagem != "" {
		t.Errorf("não esperava mensagem, obteve %q", resultado.Mensagem)
	}
}

func TestExecutarOuGatear_AcaoComGate_NuncaChegaAExecutar(t *testing.T) {
	executou := false
	resultado := ExecutarOuGatear(AcaoProposta{
		Tipo:            "y",
		RequerAprovacao: true,
		Executar: func() string {
			executou = true
			return "nunca deveria rodar"
		},
	})

	if resultado.Status != "aguardando_aprovacao" {
		t.Errorf("esperava status=aguardando_aprovacao, obteve %q", resultado.Status)
	}
	if resultado.Resultado != "" {
		t.Errorf("não esperava resultado, obteve %q", resultado.Resultado)
	}
	if executou {
		t.Error("Executar não deveria ter rodado quando RequerAprovacao=true")
	}
}
